### PR TITLE
Add SBF tests for `sol_get_sysvar`

### PR DESCRIPTION
**Pull request by buffalojoec**
#### Problem
We have some test coverage for the `sol_get_sysvar` syscall with mocks, including some tests for the custom APIs for `SlotHashes` and `StakeHistory`, but it would be great to have some live program tests within a BPF context to make sure the syscall works as expected in on-chain programs.

#### Summary of Changes
Add SBF tests under `programs/sbf` for the `sol_get_sysvar` syscall!

### CODE REVIEW

File: programs/sbf/rust/sysvar/src/lib.rs [hunk 1]
```diff
@@ -10,114 +10,206 @@ use solana_program::{
...
+        stake_history::{StakeHistory, StakeHistorySysvar},
+        Sysvar, SysvarId,
     },
 };
 
+// Adapted from `solana_program::sysvar::get_sysvar` (private).
+#[cfg(target_os = "solana")]
+fn sol_get_sysvar_handler<T>(dst: &mut [u8], offset: u64, length: u64) -> Result<(), ProgramError>
+where
+    T: Sysvar + SysvarId,
```
Comment (joncinque) [inline programs/sbf/rust/sysvar/src/lib.rs]: nit: the `Sysvar` trait requires `SysvarId`, so there's no need to specify both
```suggestion
    T: Sysvar,
```

File: programs/sbf/rust/sysvar/src/lib.rs [hunk 2]
```diff
@@ -10,114 +10,206 @@ use solana_program::{
...
+    match result {
+        solana_program::entrypoint::SUCCESS => Ok(()),
+        e => Err(e.into()),
+    }
+}
+
+// Double-helper arrangement is easier to write to a mutable slice.
+fn sol_get_sysvar<T>() -> Result<T, ProgramError>
+where
+    T: Sysvar + SysvarId,
```
Comment (joncinque) [inline programs/sbf/rust/sysvar/src/lib.rs]: nit: same here, can just do `Sysvar`
```suggestion
    T: Sysvar,
```

File: programs/sbf/tests/sysvar.rs [hunk 3]
```diff
@@ -40,6 +43,28 @@ fn test_sysvar_syscalls() {
         ..epoch_rewards::EpochRewards::default()
     };
     bank.set_sysvar_for_tests(&epoch_rewards);
+
+    // `StakeHistoryGetEntry` API assumes `MAX_ENTRIES` are present in the
+    // sysvar data.
```
Comment (joncinque) [inline programs/sbf/tests/sysvar.rs]: I may be misunderstanding the comment, but the call does work with fewer than `MAX_ENTRIES`. It just returns `None` if there are no entries: https://github.com/anza-xyz/agave/blob/7be87d95276fcb5164df9bbc068fa9497c7d10c1/sdk/program/src/sysvar/stake_history.rs#L78

**Review by joncinque — DISMISSED** [review]
Looks good! Just some tiny things, but this can go in

File: programs/sbf/tests/sysvar.rs [hunk 4]
```diff
@@ -40,6 +43,28 @@ fn test_sysvar_syscalls() {
...
+    //
+    // This isn't going to align with the bank's epoch, but it doesn't matter
+    // for the on-chain query, as long as there is data to retrieve.
```
Comment (2501babe) [inline programs/sbf/tests/sysvar.rs]: just to clarify, this code:
```rust
        let oldest_historical_epoch = current_epoch.saturating_sub(MAX_ENTRIES as u64);

        // target epoch is old enough to have fallen off history; presume fully active/deactive
        if target_epoch < oldest_historical_epoch {
            return None;
        }
```

doesnt assume there are `MAX_ENTRIES` in the data, it just uses the value to clamp the oldest epoch to 0. if stake history has fewer than 512 epochs of entries, it never reads before the first entry

Comment (buffalojoec) [reply programs/sbf/tests/sysvar.rs]: Ahh, I see. thanks! Looks like I misinterpreted the implementation. Updated my test!

File: programs/sbf/tests/sysvar.rs (hunk 3 above)
Comment (buffalojoec) [reply programs/sbf/tests/sysvar.rs]: True, updated. https://github.com/anza-xyz/agave/pull/3288#discussion_r1815838717

**Review by joncinque — APPROVED** [review]

**Pull request merged**